crates/bevy_math/src/curve/easing.rs

```diff
@@ -0,0 +1,328 @@
+
+use crate::{
+    ops::{self, FloatPow},
+    prelude::CubicSegment,
+    VectorSpace,
+};
+
+use super::{Curve, FunctionCurve, Interval};
+
+/// A trait for [`Curves`] that map the [unit interval] to some other values. These kinds of curves
+/// are used to create a transition between two values. Easing curves are most commonly known from
+/// [CSS animations] but are also widely used in other fields.
+///
+/// [unit interval]: `Interval::UNIT`
+/// [`Curves`]: `Curve`
+/// [CSS animations]: https://developer.mozilla.org/en-US/docs/Web/CSS/easing-function
+pub trait Easing<T>: Curve<T> {}
+impl<T: VectorSpace, C: Curve<f32>> Easing<T> for EasingCurve<T, C> {}
+impl<T: VectorSpace> Easing<T> for LinearCurve<T> {}
+impl Easing<f32> for CubicBezierCurve {}
+impl Easing<f32> for StepCurve {}
+impl Easing<f32> for ElasticCurve {}
+
+/// A [`Curve`] that is defined by a `start` and an `end` point, together with an [`EasingCurve`]
+/// to interpolate between the values over the [unit interval].
+///
+/// [unit interval]: `Interval::UNIT`
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
+pub struct EasingCurve<T, E>
+where
+    T: VectorSpace,
+    E: Curve<f32>,
+{
+    start: T,
+    end: T,
+    easing: E,
+}
+
+impl<T, E> Curve<T> for EasingCurve<T, E>
+where
+    T: VectorSpace,
+    E: Curve<f32>,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        Interval::UNIT
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, t: f32) -> T {
+        let domain = self.easing.domain();
+        let t = domain.start().lerp(domain.end(), t);
+        self.start.lerp(self.end, self.easing.sample_unchecked(t))
+    }
+}
+
+impl<T, E> EasingCurve<T, E>
+where
+    T: VectorSpace,
+    E: Curve<f32>,
+{
+    /// Create a new [`EasingCurve`] over the [unit interval] which transitions between a `start`
+    /// and an `end` value based on the provided [`Curve<f32>`] curve.
+    ///
+    /// If the input curve's domain is not the unit interval, then the [`EasingCurve`] will ensure
+    /// that this invariant is guaranteed by internally [reparametrizing] the curve to the unit
+    /// interval.
+    ///
+    /// [`Curve<f32>`]: `Curve`
+    /// [unit interval]: `Interval::UNIT`
+    /// [reparametrizing]: `Curve::reparametrize`
+    pub fn new(start: T, end: T, easing: E) -> Result<Self, EasingCurveError> {
+        easing
+            .domain()
+            .is_bounded()
+            .then_some(Self { start, end, easing })
+            .ok_or(EasingCurveError)
+    }
+}
+
+impl EasingCurve<f32, FunctionCurve<f32, fn(f32) -> f32>> {
+    /// A [`Curve`] mapping the [unit interval] to itself.
+    ///
+    /// Quadratic easing functions can have exactly one critical point. This is a point on the function
+    /// such that `f′(x) = 0`. This means that there won't be any sudden jumps at this point leading to
```

Comment:
Nit: The `Curve` API generally uses `t` for the time domain instead of `x`, but this notation might be a bit more familiar to people in this context, so I dunno. (This goes for everywhere)